### ISSUE TITLE
Pylint catches missing method in some situations but not in others

### ISSUE THREAD

**Issue by martimlobao** [PyCQA/pylint#5635]
### Bug description

Pylint ignores a `no-member` error when invoking a non-existing method from a Spark Context class if the class is returned from another method, but correctly recognizes the method does not exist if I call it on a `SparkContext` object directly.

I don't know if this is an issue with spark/pyspark specifically or if it's a more general issue, or if it's something that I should be doing differently.

```python
from pyspark import SparkContext
from pyspark.sql import SparkSession


class SparkInit:
    def __init__(self):
        self.active_spark_session = None

    @property
    def spark(self) -> SparkSession:
        if self.active_spark_session is None:
            self.active_spark_session = SparkSession.builder.config().getOrCreate()
        return self.active_spark_session

    @property
    def context(self) -> SparkContext:
        return self.spark.sparkContext


def main():
    # this passes
    spark_context = SparkInit().context
    spark_context.blabla()

    # this fails
    spark_context = SparkContext()
    spark_context.blabla()
```

### Configuration

```ini
(no .pylintrc configuration)
```

### Command used

```shell
python -m pylint -E example.py
```

### Pylint output

```shell
************* Module example
example.py:27:4: E1101: Instance of 'SparkContext' has no 'blabla' member (no-member)
```

### Expected behavior

Pylint should catch missing methods in both cases:

```
************* Module example
example.py:23:4: E1101: Instance of 'SparkContext' has no 'blabla' member (no-member)
example.py:27:4: E1101: Instance of 'SparkContext' has no 'blabla' member (no-member)
```

### Pylint version

```shell
pylint 2.12.2
astroid 2.9.2
Python 3.9.7 (default, Oct 29 2021, 15:18:08) 
[Clang 13.0.0 (clang-1300.0.29.3)]
```

### OS / Environment

macOS 12.1 (21C52)

### Additional dependencies

astroid==2.9.2
isort==5.10.1
lazy-object-proxy==1.7.1
mccabe==0.6.1
platformdirs==2.4.1
py4j==0.10.9.2
pylint==2.12.2
pyspark==3.2.0
toml==0.10.2
typing_extensions==4.0.1
wrapt==1.13.3

**Comment by Pierre-Sassoulas** [comment]
Thank you for opening the issue. Without looking too much into it, it's probable that our internal ast, astroid, do not understand this code construct and need better parsing capability, or more reasonably a hard coded tip for spark (we call that an astroid brain).
